zkevm-circuits/src/evm_circuit/execution/push.rs

```diff
@@ -35,10 +40,33 @@ impl<F: Field> ExecutionGadget<F> for PushGadget<F> {
 
     fn configure(cb: &mut EVMConstraintBuilder<F>) -> Self {
         let opcode = cb.query_cell();
+        let code_size = cb.query_cell();
+        let code_size_left = code_size.expr() - cb.curr.state.program_counter.expr();
 
         let value = cb.query_word32();
-        // Query selectors for each opcode_lookup
-        let selectors = array_init(|_| cb.query_bool());
+        // Query selectors for each opcode_lookup to know whether the current byte needs to be
+        // pushed
+        let is_pushed = array_init(|_| cb.query_bool());
+
+        // Query selectors for each opcode_lookup to know whether the current byte is actually
+        // padding
+        let is_padding = array_init(|_| cb.query_bool());
+
+        // Fetch the bytecode length from the bytecode table.
+        let code_hash = cb.curr.state.code_hash.clone();
+        cb.bytecode_length(code_hash.to_word(), code_size.expr());
+
+        // Deduce the number of bytes to push.
+        let num_pushed = opcode.expr() - OpcodeId::PUSH1.as_u64().expr() + 1.expr();
+
+        // If the number of bytes that needs to be pushed is greater
+        // than the size of the bytecode left, then we are out of range.
+        let is_out_of_range: LtGadget<F, N_BYTES_PROGRAM_COUNTER> =
+            LtGadget::construct(cb, code_size_left.clone(), num_pushed.clone());
+
+        // Expected number of padding
+        let num_padding: halo2_proofs::plonk::Expression<F> = 32.expr() - num_pushed.clone()
```

Comment:
In the specs this is `num_padding = is_out_of_bound * (num_pushed - code_length_left)`.  The `+1` in this PR probably comes from the difference of how `code_size_left` is calculated.

But what about the `32 - num_pushed`?

Reply:
Perhaps the terminology is not the best (as well as the consistency sorry!).
Basically, I called by padding here anything that is not pushed, so that corresponds to the 32 - n bytes that are not pushed. Likewise, when we are out of bound we have `(num_pushed - code_length_left)` extra bytes that are set to 0.

The +1 comes exactly from this reason, I made a mistake before and will fix them both.

Reply:
To add a bit more:  We want to check that either the bytes that are pushed are in the bytecode table and the non pushed ones are set to 0s. Finally we check that the number of byte (potentially) pushed is indeed n (`Sum is_pushed = n`).

The number of byte we actually push is not n but `n - is_out_of_bound * (num_pushed - code_length_left)`. We could thus change the previous constraint by `Sum (is_pushed * (1 - is padding)) = n - is_out_of_bound * (num_pushed - code_length_left)`. The problem I see with such equation is that the 1) identity has a higher degree, and 2) you could still set is_pushed to 1 when is_padding = 1. Not major problems, but it's not very robust.
Hence, I decided to keep the check on "n" (not the number of bytes pushed), `Sum is_pushed = n`, as well as add an extra check that the `Sum is_padding = num_padding`.

Reply:
... I see a flaw in this reasoning now when adding the `32 - n`, in `num_padding` so I will remove it!

Reply:
The flaw (if anyone interested): is padding because it is the concatenation of both the `32-n` and `(num_pushed - code_length_left)` bool has the following form: 11... 000... 111  where the first 1s should represent the out of bound padding and the last the normal 32-n.  Because we do not enforce the size of each sub list of 1s, one could move some 1s from one sublist to the others, hence we would not perform some of the bytecode table lookups.